Replace `IsSubType` check with `IsSameTypeKind` check where necessary

SupunS opened the pull request:
Closes #1031 

## Description

- Introduce a method to check whether a given type 'belongs' to the same kind as a given type. 
  - Same as IsSubType, but doesn't treat `Never` as 'not' belongs to the same kind.
- Replace the usages where strictly 'same kind types' check is needed.

______

<!-- Complete: -->

- [x] Targeted PR against `master` branch
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work
- [x] Code follows the [standards mentioned here](https://github.com/onflow/cadence/blob/master/CONTRIBUTING.md#styleguides)
- [x] Updated relevant documentation 
- [x] Re-reviewed `Files changed` in the Github PR explorer
- [x] Added appropriate labels

turbolent commented on runtime/sema/checker.go:
```diff
@@ -608,16 +601,9 @@ func (checker *Checker) checkTypeCompatibility(expression ast.Expression, valueT
-			break
-		}
-
 		valueTypeOK := CheckFixedPointLiteral(typedExpression, valueType, checker.report)
```
The check above which was removed made this line conditional. Should this line now maybe be moved below new line 606, into the body of the if-statement?

SupunS replied:
It wouldn't change the checker results, but yeah, can avoid additional computation if the condition fails. 
Refactored: e6be2687902bd3ea6cedefa01e6a36b260f4e6ae

turbolent reviewed (commented):
Great work!

Do we have a way of auditing the rest of the code for cases where `IsSameTypeKind` should be used instead of `IsSubType`? Also, maybe we can document a guideline for future code?

SupunS commented:
> Do we have a way of auditing the rest of the code for cases where IsSameTypeKind should be used instead of IsSubType?

I searched all the usages and replaced them where necessary. Unfortunately, this has to be done manually, can't think of a way to automate this/ audit with a test case. The general rule is:
- `IsSameType()`  - To check the assignability. e.g: Is argument type `T` is a sub-type of parameter type `R`. This is the more frequent use-case.
- `IsSAmeYTypeKind()` - To check if a type strictly belongs to a certain category. e.g: Is the expression type `T` is any of the integer types, but nothing else. Another way to check is, asking the question "if the `subType` is `Never`, should the check still pass?"

> Also, maybe we can document a guideline for future code?

I had  already added some comments to the newly added mehtod. Also added a note on the old `IsSameType()` method for the usage: 1b324ec1b79e3ce5306788eac0e97ad4420273a9

turbolent commented:
@SupunS Nice! Could we keep this comment/guideline somewhere in the codebase? Maybe as a comment somewhere, or in `docs`?

turbolent reviewed (approved):
Great work! 👏

turbolent commented on runtime/sema/type.go:
```diff
@@ -4708,9 +4708,17 @@ func (t *AddressType) GetMembers() map[string]MemberResolver {
-// However, to check if a type *strictly* belongs to a certain category, then
-// consider using `IsSameTypeKind` method. e.g: "Is type `T` an Integer type?". Using
-// this method for the later use-case may produce incorrect results.
+// However, to check if a type *strictly* belongs to a certain category, then consider
+// using `IsSameTypeKind` method. e.g: "Is type `T` an Integer type?". Using this method
+// for the later use-case may produce incorrect results.
+//   * IsSubType()      - To check the assignability. e.g: Is argument type T is a sub-type
+//                        of parameter type R. This is the more frequent use-case.
+//   * IsSameTypeKind() - To check if a type strictly belongs to a certain category. e.g: Is the
+//                        expression type T is any of the integer types, but nothing else.
+//                        Another way to check is, asking the question of "if the subType is Never,
+//                        should the check still pass?". A common code-smell for potential incorrect
+//                        usage is, using IsSubType() method with a constant/pre-defined superType.
+//                        e.g: IsSubType(<<someType>>, FixedPointType)
```
✨ 👏
